### PR TITLE
Add rekognition-service

### PR DESCRIPTION
**Pull request by Manuelacosta98**
Hi great library, i am trying to add rekognition service to async-aws, i encounter a problem with the method  SearchFacesByImage  the problem is the next:

`In RestJsonSerializer.php line 158:
                                     
  Type float is not yet implemented `

I am unable to track the issue, i am a newbie in OSS but want to get this done, Thanks for the help i will start adding the test in the meanwhile

### CODE REVIEW

**Comment by Manuelacosta98** [comment]
> Hi great library, i am trying to add rekognition service to async-aws, i encounter a problem with the method SearchFacesByImage the problem is the next:
> 
> `In RestJsonSerializer.php line 158:
> 
> Type float is not yet implemented `
> 
> I am unable to track the issue, i am a newbie in OSS but want to get this done, Thanks for the help i will start adding the test in the meanwhile

The problem seams to be in the function dumpArrayElement in RestJsonSerializer

` 
        switch ($shape->getType()) {
            case 'string':
            case 'integer':
            case 'float':
            case 'long':
                return $this->dumpArrayScalar($output, $input, $contextProperty, $shape);
            case 'boolean':
                return $this->dumpArrayBoolean($output, $input, $shape);
            case 'timestamp':
                return $this->dumpArrayTimestamp($output, $input, $shape);
            case 'blob':
                return $this->dumpArrayBlob($output, $input, $shape, $isRequired);
        }`

I add the float type to the scalar value case in the function.

So if someone can take a look of this i will be testing it soon.

**Comment by jderusse** [comment]
Thanks for your contribution. Looks like your fix is OK. We did not implemented `FLOAT` in Json because we didn't had use case for it.
Could you confirm that passing a float in `FaceMatchThreshold` to `SearchFacesByImage` is accepted by AWS?
and calling `getSearchedFaceConfidence` is a valid float in the repsonse?

You'll need to rebase your PR (fetch last version) and run `generate --all` to fix `sts`/`sqs` coding standard
And implements tests that have been partially generated.

**Comment by Manuelacosta98** [comment]
> Thanks for your contribution. Looks like your fix is OK. We did not implemented `FLOAT` in Json because we didn't had use case for it.
> Could you confirm that passing a float in `FaceMatchThreshold` to `SearchFacesByImage` is accepted by AWS?
> and calling `getSearchedFaceConfidence` is a valid float in the repsonse?
> 
> You'll need to rebase your PR (fetch last version) and run `generate --all` to fix `sts`/`sqs` coding standard
> And implements tests that have been partially generated.

Hi @jderusse thanks for replying i will give it a try today and answer. i dint get the rebase part you got a tag or branch on witch i need to rebase? 

Sorry i fork the repo from master i am new to these.

**Comment by jderusse** [comment]
see the warning "This branch has conflicts that must be resolved"  in the GitHub interface.
it means that somebody else pushed in master (another PR) and your fork is not up to date.
Given the conflicted files are not related I think running `generate --all` should fix the issue and you don't have to rebase.

but if you are curious and want more information about rebase and fork, have a look to https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request

**Comment by Manuelacosta98** [comment]
> see the warning "This branch has conflicts that must be resolved" in the GitHub interface.
> it means that somebody else pushed in master (another PR) and your fork is not up to date.
> Given the conflicted files are not related I think running `generate --all` should fix the issue and you don't have to rebase.
> 
> but if you are curious and want more information about rebase and fork, have a look to https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request

Hi @jderusse sorry for the time past i got two problems the first is a question who do i run the test only in the new service? , and the other is that i try to require my fork repo in a project to test it and i encounter with this error 
Screenshot from 2020-08-13 10-46-18

is there a way to include via composer the recognition client from my fork repo? 

Thanks a lot

**Comment by Manuelacosta98** [comment]
> > see the warning "This branch has conflicts that must be resolved" in the GitHub interface.
> > it means that somebody else pushed in master (another PR) and your fork is not up to date.
> > Given the conflicted files are not related I think running `generate --all` should fix the issue and you don't have to rebase.
> > but if you are curious and want more information about rebase and fork, have a look to https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request
> 
> Hi @jderusse sorry for the time past i got two problems the first is a question who do i run the test only in the new service? , and the other is that i try to require my fork repo in a project to test it and i encounter with this error
> Screenshot from 2020-08-13 10-46-18
> 
> is there a way to include via composer the recognition client from my fork repo?
> 
> Thanks a lot

When i try to run the individual test with my global phpunit i get these error:

 "PHP Fatal error:  Uncaught Error: Class 'AsyncAws\Core\Test\TestCase' not found"

I already run the composer install , theres another step to take before testing or only can be done with the docker makefile?

**Comment by Manuelacosta98** [comment]
Ok i found out how is done, i am fixing the falling test, Thanks

**Comment by Manuelacosta98** [comment]
Just keeping the updates i implemented the request unit test in the new service an also added some basic documentation to pass those test as well, tomorrow i will implement the test in the unit response side, i do not have knowledge in docker container i think some of these test can only be implemented that way i am not quite sure.

Thanks a lot for the follow up.

File: src/Service/Rekognition/.github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+jobs:
+
+  build:
+    name: Build
+    runs-on: ubuntu-latest
+    strategy:
+      max-parallel: 10
+      matrix:
+        php: ['7.2', '7.3', '7.4']
```
Comment (jderusse) [inline src/Service/Rekognition/.github/workflows/ci.yml:16]: missing php8? 

/cc @Nyholm

File: composer.json [hunk 2]
```diff
@@ -67,7 +67,8 @@
             "AsyncAws\\Illuminate\\Mail\\": "src/Integration/Laravel/Mail/src",
             "AsyncAws\\Illuminate\\Queue\\": "src/Integration/Laravel/Queue/src",
             "AsyncAws\\SimpleS3\\": "src/Integration/Aws/SimpleS3/src",
-            "AsyncAws\\Symfony\\Bundle\\": "src/Integration/Symfony/Bundle/src"
+            "AsyncAws\\Symfony\\Bundle\\": "src/Integration/Symfony/Bundle/src",
+            "AsyncAws\\Rekognition\\": "src/Service/Rekognition/src"
```
Comment (Nyholm) [inline composer.json:71]: Put it just over S3 to keep this in alphabetical order and next to the other clients.

File: composer.json [hunk 3]
```diff
@@ -96,7 +97,8 @@
             "AsyncAws\\Illuminate\\Mail\\Tests\\": "src/Integration/Laravel/Mail/tests",
             "AsyncAws\\Illuminate\\Queue\\Tests\\": "src/Integration/Laravel/Queue/tests",
             "AsyncAws\\SimpleS3\\Tests\\": "src/Integration/Aws/SimpleS3/tests",
-            "AsyncAws\\Symfony\\Bundle\\Tests\\": "src/Integration/Symfony/Bundle/tests"
+            "AsyncAws\\Symfony\\Bundle\\Tests\\": "src/Integration/Symfony/Bundle/tests",
+            "AsyncAws\\Rekognition\\Tests\\": "src/Service/Rekognition/tests"
```
Comment (Nyholm) [inline composer.json:101]: Same here

File: couscous.yml [hunk 4]
```diff
@@ -67,6 +67,10 @@ menu:
             sts:
                 text: STS
                 url: /clients/sts.html
+            rekognition:
```
Comment (Nyholm) [inline couscous.yml:70]: Same here, Just above S3

File: docs/clients/index.md [hunk 5]
```diff
@@ -149,3 +149,4 @@ for more information.
 | [SQS](./sqs.md)                             | [async-aws/sqs](https://packagist.org/packages/async-aws/sqs)
 | [SSM](./ssm.md)                             | [async-aws/ssm](https://packagist.org/packages/async-aws/ssm)
 | [STS](./sts.md)                             | [async-aws/core](https://packagist.org/packages/async-aws/core)
+| [Rekognition](./rekognition.md)             | [async-aws/rekognition](https://packagist.org/packages/async-aws/rekognition)
```
Comment (Nyholm) [inline docs/clients/index.md:152]: Same

File: docs/clients/rekognition.md [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+$rekognition = new RekognitionClient();
+
+$request = new CreateCollectionRequest([
+    "collectionId" => "testId",
+    "region" => "us-west-2"
+]);
+
+$response = $rekognition->createCollection($request);
+
+$response->resolve();
```
Comment (Nyholm) [inline docs/clients/rekognition.md:29]: This line is not needed.

File: docs/clients/rekognition.md [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+When you want to search a face inside a collection with an image you will use the method searchFacesByImage.
+
+```php
+use AsyncAws\Rekognition\RekognitionClient;
+use AsyncAws\Rekognition\Input\searchFacesByImageRequest;
+
+$rekognition = new RekognitionClient();
+
+$addFacesRequest = new searchFacesByImageRequest(
```
Comment (Nyholm) [inline docs/clients/rekognition.md:66]:
```suggestion
$addFacesRequest = new SearchFacesByImageRequest(
```

File: src/Core/src/AwsClientFactory.php [hunk 8]
```diff
@@ -258,4 +259,17 @@ public function cognitoIdentityProvider(): CognitoIdentityProviderClient
 
         return $this->serviceCache[__METHOD__];
     }
+
+    public function rekognition(): RekognitionClient
```
Comment (Nyholm) [inline src/Core/src/AwsClientFactory.php:263]: Could you put this just over S3?

File: src/Integration/Symfony/Bundle/src/DependencyInjection/AwsPackagesProvider.php [hunk 9]
```diff
@@ -69,6 +69,10 @@ public static function getAllServices(): array
                 'class' => \AsyncAws\Core\Sts\StsClient::class,
                 'package' => 'async-aws/core',
             ],
+            'rekognition' => [
```
Comment (Nyholm) [inline src/Integration/Symfony/Bundle/src/DependencyInjection/AwsPackagesProvider.php:72]: Same

File: src/Integration/Symfony/Bundle/src/DependencyInjection/AwsPackagesProvider.php [hunk 10]
```diff
@@ -69,6 +69,10 @@ public static function getAllServices(): array
...
+                'class' => \AsyncAws\Rekognition\RekognitionClient::class,
+                'package' => 'async-aws/Rekognition',
```
Comment (Nyholm) [inline src/Integration/Symfony/Bundle/src/DependencyInjection/AwsPackagesProvider.php:74]:
```suggestion
                'package' => 'async-aws/rekognition',
```

File: src/Service/Rekognition/tests/Unit/Result/DeleteProjectResponseTest.php [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+use AsyncAws\Rekognition\RekognitionClient;
+use AsyncAws\Rekognition\Result\DeleteProjectResponse;
+use Psr\Log\NullLogger;
+use Symfony\Component\HttpClient\MockHttpClient;
+
+class DeleteProjectResponseTest extends TestCase
+{
+    public function testDeleteProjectResponse(): void
+    {
+        self::fail('Not implemented');
```
Comment (Nyholm) [inline src/Service/Rekognition/tests/Unit/Result/DeleteProjectResponseTest.php:18]: Make sure we complete all these tests

File: README.md [hunk 12]
```diff
@@ -17,13 +17,14 @@ See full documentation on [https://async-aws.com](https://async-aws.com).
...
-| [async-aws/iam](https://github.com/async-aws/iam)                                             | [![Latest Stable Version](https://poser.pugx.org/async-aws/iam/v/stable)](https://packagist.org/packages/async-aws/iam)                                             [![Total Downloads](https://poser.pugx.org/async-aws/iam/downloads)](https://packagist.org/packages/async-aws/iam)                           | [![](https://github.com/async-aws/event-bridge/workflows/BC%20Check/badge.svg?branch=master)](https://github.com/async-aws/event-bridge/actions)
+| [async-aws/iam](https://github.com/async-aws/iam)                                             | [![Latest Stable Version](https://poser.pugx.org/async-aws/iam/v/stable)](https://packagist.org/packages/async-aws/iam)                                             [![Total Downloads](https://poser.pugx.org/async-aws/iam/downloads)](https://packagist.org/packages/async-aws/iam)                                             | [![](https://github.com/async-aws/event-bridge/workflows/BC%20Check/badge.svg?branch=master)](https://github.com/async-aws/event-bridge/actions)
```
Comment (Nyholm) [inline README.md:20]: We should not change this line.

I dont really see what's changed to be honest.

File: src/Service/Rekognition/.github/workflows/branch_alias.yml [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+          ARR[1]=$((${ARR[1]}+1))
+          echo ::set-output name=alias::${ARR[0]}.${ARR[1]}
+
+      - name: Checkout main repo
+        run: |
+          git clone --branch master https://${{ secrets.BOT_GITHUB_TOKEN }}:x-oauth-basic@github.com/async-aws/aws aws
+
+      - name: Update branch alias
+        run: |
+          cd aws/src/Service/Foobar
```
Comment (Nyholm) [inline src/Service/Rekognition/.github/workflows/branch_alias.yml:34]:
```suggestion
          cd aws/src/Service/Rekognition
```

**Review by Nyholm — APPROVED** [review]
Thank you

File: src/Service/Rekognition/.github/workflows/branch_alias.yml (hunk 13 above)
Comment (Nyholm) [reply src/Service/Rekognition/.github/workflows/branch_alias.yml:34]: This is why the rootTest fails.

**Comment by Nyholm** [comment]
The new psalm failures are false positive. 

Psalm says: "We cannot trust the data coming form the server". Which we really can since we have a contract.

**Comment by jderusse** [comment]
thank you @Nyholm for your help.

**Comment by Nyholm** [comment]
Thank you @jderusse and @Manuelacosta98 

https://packagist.org/packages/async-aws/rekognition

**Comment by Manuelacosta98** [comment]
Thanks for all i got pretty sick sorry about that

**Comment by Nyholm** [comment]
We are happy to help. 

I hope you have recovered and feel better now. Your health is always more important than some open source work.

**Pull request merged**